docs/sources/design-documents/2021-01-Ordering-Constraint-Removal.md

```diff
@@ -0,0 +1,197 @@
+
+```
+Figure 4
+
+     chunk1
+-------------------
+                chunk2
+         ---------------------
+         query range requiring both
+         ----------
+                             query range requiring chunk2 only
+                             -----------
+ts0     ts1      ts2       ts3        ts4 (not in any block)
+------------------------------
+|        |        |          |
+|        |        |          |
+|        |        |          |
+|        |        |          |
+------------------------------
+
+```
+
+The performance losses against the current approach includes:
+
+1) Appending a log line is now performed in logarithmic time instead of amortized (due to array resizing) constant time.
+2) Blocks may contain overlapping data (although ordering is still guaranteed within each block).
+3) Head block scans are now `O(n log(n))` instead of `O(n)`
+
+### Flushing & Chunk Creation
+
+Loki regularly combines multiple blocks into a chunk and "flushes" it to storage. In order to ensure that reads over flushed chunks remain as performant as possible, we will re-order a possibly-overlapping set of blocks into a set of blocks that maintain monotonically increasing order between them. From the perspective of the rest of Loki’s components (queriers/rulers fetching chunks from storage), nothing has changed.
+
+**Note: In the case that data for a stream is ingested in order, this is effectively a no-op, making it well optimized for in-order writes (which is both the requirement and default in Loki currently). Thus, this should have little performance impact on ordered data while enabling Loki to ingest unordered data.**
+
+#### Chunk Durations
+
+When `--validation.reject-old-samples` is enabled, Loki accepts incoming timestamps within the range
+```
+[now() - `--validation.reject-old-samples.max-age`, now() + `--validation.create-grace-period`]
+```
+For most of our clusters, this would mean the range of acceptable data is one week long. In contrast, our max chunk age is `2h`. Allowing unordered writes would mean that ingesters would willingly receive data for 168h, or up to 84 distinct chunk lengths. This presents a problem: a malicious user could be writing to many (84 in this case) distinct chunks simultaneously, flooding Loki with underutilized chunks which bloat the index.
+
+In order to mitigate this, there are a few options (not mutually exclusive):
+1) Lower the valid acceptance range
+2) Create an _active_ validity window, such as `[most_recent_sample-max_chunk_age, now() + creation_grace_period]`.
+
+The first option is simple, already available, and likely somewhat reasonable.
+The second is simple to implement and an effective way to ensure Loki can ingest unordered logs but maintain a sliding validity window. I expect this to cover nearly all reasonable use cases and effectively mitigate bad actors.
```

Comment:
We should do that. If we don't restraint the boundary of a chunks, it will appear too often for queries and will lead to bad performance.

As of today the only way to have overlapping time between chunk of the same stream is replication. I believe this won't be true anymore. (Not a big problem just a statement.)

Reply:
Yes, I fully intend to use the second strategy here to slowly cinch the validity window. This will allow relatively out of order writes while preventing bad actors from writing samples i.e. one week apart over and over.